### DISCUSSION TITLE
7.1 tweaking sql query issues

### DISCUSSION THREAD

**Discussion by khartman-nj** [ua-gist604b-f24/syllabus#50] [General]
I have a bug or something wrong with my code I have done exactly what the video suggests and still I get no returns. I put my code below and suggestions on were I went wrong are greatly appreciated
![Screenshot 2024-12-07 075438](https://github.com/user-attachments/assets/81d4e199-c39b-486d-8b91-a061b797efc0)
 
# uvicorn main:app --reload
from starlette.responses import FileResponse 

from fastapi import FastAPI
from databases import Database
from sqlalchemy import create_engine 

app = FastAPI() 

DATABASE_URL = "postgresql://postgres:postgres@localhost:5432/hawaii"

database = Database(DATABASE_URL)

engine = create_engine(DATABASE_URL)  # Access the DB Engine

@app.get("/")
async def read_index():
    return FileResponse('static/index.html')

# Add "search" here
@app.get("/search")
async def search(q: str):
    # This is the actual query. You may want to work out the SQL statement in a separate `psql` or other database session.
    # The start SQL will give you everything matching `place` exactly which isn't quite what you need. 
    # See the instructions for guidance on how to update this for this jquery autocomplete
    query = '''
SELECT 
  name as label, 
  name
FROM 
  import.osm_places 
WHERE 
  name like :name
'''
   # This adds a `%` to the end of the `place` parameter, changing, e.g. "Hono" to "Hono%".
    place = "{}{}".format(q,'%')
    rows = await database.fetch_all(query=query, values={"name": place})
    # For our pursposes we want to save the results in a special json format with two keys: `label` and `value`.
    # That's because jquery `autocomplete` will work if follow this format convention for our JSON output.
    
    results = []
    for row in rows:
        results.append({'label': row[0], 'value': row[1]})
    return results
# Add "find_coffee" here

@app.on_event("startup")
async def startup():
    await database.connect()

@app.on_event("shutdown")
async def shutdown():
    await database.disconnect()

**Comment by vmwielgus** [comment]
I see you put your curl commands in the python3 terminal. In the terminal window navigate to the right side of the screen and click the '+' icon to open a new terminal. It will probably be called 'bash' or something like that. Then run your curl commands in the new terminal instead of the python3 one. I just remember there were some commands that had to go into the python3 terminal and others that had to go into a different terminal, so just make sure you switch between the two at the right times in the tutorial.

**Comment by khartman-nj** [comment]
Hello I re did my process and still came to the same issue whenever I put in the Curl statement I get no results 
![Screenshot 2024-12-07 154920](https://github.com/user-attachments/assets/aa91a7b8-07ef-4c99-bcd6-d50a038cc55f)

![Screenshot 2024-12-07 154837](https://github.com/user-attachments/assets/cd0fe2d9-6a3a-483e-adf9-8af94a57240a)

**Comment by wtb4rtr** [comment]
Hi, I think you need to add the "Upper" qualifier in your main.py file in order to get the query to select Honolulu when it isnt capitalized
![7-1](https://github.com/user-attachments/assets/633eca4d-99d2-42ab-bf17-f040bc11d270)

It needs to be:
WHERE 
  upper(name) like :name
'''
     # This adds a `%` to the end of the `place` parameter, changing, e.g. "Hono" to "Hono%".
    place = "{}{}".format(q,'%').upper()

I think if you make these corrections to you main.py file you should get a result.  Good luck
Bill Brown

**Comment by khartman-nj** [comment]
<html>

<head>
  <title>Place Finder</title>
  <base target="_top">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>OSM PLace Searcher</title>

  <link rel="stylesheet" href="https://unpkg.com/leaflet@1.9.3/dist/leaflet.css"
    integrity="sha256-kLaT2GOSpHechhsozzB+flnD+zUyjE2LlfWPgU04xyI=" crossorigin="" />
  <script src="https://unpkg.com/leaflet@1.9.3/dist/leaflet.js"
    integrity="sha256-WBkoXOwTeyKclOHuWtc+i2uENFpDZ9YPdf5Hf+D7ewM=" crossorigin=""></script>

  // Add JQuery CSS and javascript here
  
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.6.0/jquery.min.js"></script>
  <script>
    $(function () {
      $("#place").autocomplete({
        source: function (request, response) {
          $.ajax({
            url: "/search",
            dataType: "json",
            data: {
              q: request.term
            },
            success: function (data) {
              response($.map(data, function (item) {
                return {
                  label: item.label,
                  value: item.value,
                  lat: item.lat,
                  lon: item.lon,
                  // Note that if we add additional items to the http response we will need to update here
                };
              }));
            }
            });
          select: function (event, ui) {
            event.preventDefault();

            var lat = ui.item.lat; // todo: this needs to be part of the `/search` response
            var lon = ui.item.lon; // todo: this needs to be part of the `/search` response

            // if the marker is not null it means we already have one that should be removed before we add another one
            if (marker != null) {
              map.removeLayer(marker)
            }
            marker = L.marker([lat, lon], { title: ui.item.label, icon: redIcon }).addTo(map);

            marker.bindPopup('<b>' + ui.item.label + '</b>').openPopup();

            // Add a button to find coffee. This button should be bound with an `onclick` action that calls a
            // javascript `findCoffee()` function that takes a longitude and latitude argument
            $("#search").html('<input type="button" value="Find coffee!" onclick="findCoffee(' + marker.getLatLng().lng + ',' + marker.getLatLng().lat + ')">');

            map.flyTo([lat, lon], 14);

            return false;
          }
        },
      });
    });
  </script>
  
  <link rel="stylesheet" href="https://ajax.googleapis.com/ajax/libs/jqueryui/1.13.2/themes/smoothness/jquery-ui.css">
  <script src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.13.2/jquery-ui.min.js"></script>

  <script>
    // Variable initialization
    // current marker
    var marker = null;

    var redIcon = new L.Icon({
      iconUrl: 'https://raw.githubusercontent.com/pointhi/leaflet-color-markers/master/img/marker-icon-2x-red.png',
      shadowUrl: 'https://cdnjs.cloudflare.com/ajax/libs/leaflet/0.7.7/images/marker-shadow.png',
      iconSize: [25, 41],
      iconAnchor: [12, 41],
      popupAnchor: [1, -34],
      shadowSize: [41, 41]
    });

    // Place search autocomplete

    // find coffee closest to a given place
    function findCoffee(lon, lat) {
      // todo: complete this
    }

    function clearResults() {
      // todo: complete this
    }
  </script>

</head>

<body>

  <div id="map" style="width: 600px; height: 400px;"></div>
  <div class="ui-widget">
    <label for="place">Enter Place Name: </label>
    <input id="place">
  </div>
  <div id="search"></div>
  <div id="result"></div>
  <div id="hide_results" style="display:none;"><input type="button" value="clear results" onclick="clearResults()">
  </div>

  <script>
    // Create a map
    var map = new L.Map('map', { center: new L.LatLng(19.5429, -155.6659), zoom: 8, attributionControl: true, zoomControl: false });
    var osmUrl = 'http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png';
    var osmAttrib = 'Map data © <a href="http://openstreetmap.org">OpenStreetMap</a> contributors';
    var osm = new L.TileLayer(osmUrl, { minZoom: 3, maxZoom: 18, attribution: osmAttrib });
    map.addLayer(osm);

    // print the lat/long on the map as the mouse moves
    let Position = L.Control.extend({
      _container: null,
      options: {
        position: 'bottomleft'
      },

      onAdd: function (map) {
        var latlng = L.DomUtil.create('div', 'mouseposition');
        this._latlng = latlng;
        return latlng;
      },

      updateHTML: function (lat, lng) {
        var latlng = lat + " " + lng;
        this._latlng.innerHTML = "LatLng: " + latlng;
      }
    });
    this.position = new Position();
    this.map.addControl(this.position);
    this.map.addEventListener('mousemove', (event) => {
      let lat = Math.round(event.latlng.lat * 100000) / 100000;
      let lng = Math.round(event.latlng.lng * 100000) / 100000;
      this.position.updateHTML(lat, lng);
    });
  </script>
</body>

</html>
I have a simple syntax error on this section of my code any suggestions on how to debug this ?
![syntax](https://github.com/user-attachments/assets/a21656da-4085-4024-9077-5e673f891921)

**Comment by khartman-nj** [comment]
![syntax](https://github.com/user-attachments/assets/5ccd8f27-e038-4f44-b57b-f93cc018af87)

**Reply by khartman-nj** [reply]
Any suggestions on how to get this debugged ?

**Comment by wtb4rtr** [comment]
You need another curly bracket and a comma before the select function.  Bill Brown
![7-1](https://github.com/user-attachments/assets/525234b4-3a11-4afc-aca3-522082c2c84c)
